tsweb/varz/varz_test.go

```diff
@@ -418,3 +420,71 @@ func TestVarzHandlerSorting(t *testing.T) {
 		}
 	}
 }
+
+func TestWriteMemestats(t *testing.T) {
+	memstats := &runtime.MemStats{
+		Alloc:        1,
+		TotalAlloc:   2,
+		Sys:          3,
+		Lookups:      4,
+		Mallocs:      5,
+		Frees:        6,
+		HeapAlloc:    7,
+		HeapSys:      8,
+		HeapIdle:     9,
+		HeapInuse:    10,
+		HeapReleased: 11,
+		HeapObjects:  12,
+		StackInuse:   13,
+		StackSys:     14,
+		MSpanInuse:   15,
+		MSpanSys:     16,
+		MCacheInuse:  17,
+		MCacheSys:    18,
+		BuckHashSys:  19,
+		GCSys:        20,
+		OtherSys:     21,
+		NextGC:       22,
+		LastGC:       23,
+		PauseTotalNs: 24,
+		// PauseNs:       [256]int64{},
+		NumGC:         26,
+		NumForcedGC:   27,
+		GCCPUFraction: 28.1,
+	}
+
+	var buf bytes.Buffer
+	writeMemstats(&buf, memstats)
+	lines := strings.Split(buf.String(), "\n")
+
+	checkFor := func(name, typ, value string) {
+		var foundType, foundValue bool
+		for _, line := range lines {
+			if line == "memstats_"+name+" "+value {
+				foundValue = true
+			}
+			if line == "# TYPE memstats_"+name+" "+typ {
+				foundType = true
+			}
+			if foundValue && foundType {
+				return
+			}
+		}
+		t.Errorf("memstats_%s foundType=%v foundValue=%v", name, foundType, foundValue)
+	}
+
+	t.Logf("memstats:\n %s", buf.String())
+
+	checkFor("heap_alloc", "gauge", "7")
+	checkFor("total_alloc", "counter", "2")
+	checkFor("sys", "gauge", "3")
+	checkFor("mallocs", "counter", "5")
+	checkFor("frees", "counter", "6")
+	checkFor("num_gc", "counter", "26")
+	checkFor("gc_cpu_fraction", "gauge", "28.1")
+
+	testing.AllocsPerRun(1000, func() {
```

Comment:
this returns the number of allocs. You have to test it against your expected value.

Reply:
Sigh. Oh well. Can optimize it later I guess.

Reply:
I switched most of the printf for write, that cut it from 30 to 17, while adding the new lines.

Reply:
You'll probably want to disable the alloc test if race `racebuild.On`